### PR TITLE
Comment on why we can't replace copy_addr instructions with moves in arc-inert blocks

### PR DESCRIPTION
**Pull request by shajrawi**
radar rdar://problem/24011383

This concludes this radar by adding a comment on why copy_addr instructions can't be turned into moves in simplifyProgramTerminationBlock()

### CODE REVIEW

**Comment by shajrawi** [comment]
@gottesmm Can you please review?

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Comment by jckarter** [comment]
Why does it have to be ARC-inert? It should always be possible to break a `copy_addr [init] [take]` into a load and store.

**Comment by gottesmm** [comment]
@jckarter That is not what is happening here. What is happening here is that we are changing copy_addr in blocks that we know are leaking to just be moves.

**Comment by jckarter** [comment]
The transformation here is always valid; `copy_addr [init] [take]` should always be equivalent to `load; store` for a loadable type. I was wondering why we were constraining the transformation only to 'ARC-inert' blocks; AFAICS this transformation should always be OK.

**Comment by gottesmm** [comment]
@shajrawi Sorry for the delay. I am reviewing this now.

File: lib/SILOptimizer/Transforms/SimplifyCFG.cpp [hunk 1]
```diff
@@ -3418,11 +3418,26 @@ bool SimplifyCFG::simplifyProgramTerminationBlock(SILBasicBlock *BB) {
   // releases at this point.
   bool Changed = false;
   llvm::SmallPtrSet<SILInstruction *, 4> InstsToRemove;
+  llvm::SmallPtrSet<SILInstruction *, 4> InstsToReplace;
   for (auto &I : *BB) {
-    if (!isa<StrongReleaseInst>(I) && !isa<UnownedReleaseInst>(I) && 
-        !isa<ReleaseValueInst>(I) && !isa<DestroyAddrInst>(I))
+    if (!isa<StrongReleaseInst>(I) && !isa<UnownedReleaseInst>(I) &&
+        !isa<ReleaseValueInst>(I) && !isa<DestroyAddrInst>(I) &&
+        !isa<CopyAddrInst>(I))
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/SimplifyCFG.cpp]: Maybe refactor this query into a helper function?

File: lib/SILOptimizer/Transforms/SimplifyCFG.cpp [hunk 2]
```diff
@@ -3418,11 +3418,26 @@ bool SimplifyCFG::simplifyProgramTerminationBlock(SILBasicBlock *BB) {
...
       continue;
     InstsToRemove.insert(&I);
+    if (isa<CopyAddrInst>(I)) {
+      InstsToReplace.insert(&I);
+    }
+  }
+
+  // Add moves for every copy_addr
+  for (auto I : InstsToReplace) {
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/SimplifyCFG.cpp]: May be good to make it clear that this is a pointer by making it an auto *

File: lib/SILOptimizer/Transforms/SimplifyCFG.cpp [hunk 3]
```diff
@@ -3418,11 +3418,26 @@ bool SimplifyCFG::simplifyProgramTerminationBlock(SILBasicBlock *BB) {
...
+    SILBuilderWithScope Builder(I);
+    auto CAI = cast<CopyAddrInst>(I);
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/SimplifyCFG.cpp]: Can you also make this an auto *.

File: lib/SILOptimizer/Transforms/SimplifyCFG.cpp [hunk 4]
```diff
@@ -3418,11 +3418,26 @@ bool SimplifyCFG::simplifyProgramTerminationBlock(SILBasicBlock *BB) {
...
+    assert(CAI && "Expected CopyAddrInst in InstsToReplace set");
+    auto Loc = CAI->getLoc();
+    auto LI = Builder.createLoad(Loc, CAI->getSrc());
```
Comment (gottesmm) [inline lib/SILOptimizer/Transforms/SimplifyCFG.cpp]: Also this please make an auto * ; ).

**Review by gottesmm — CHANGES_REQUESTED** [review]
I think we can make this even more aggressive. Given that we are always going to leak in these blocks, we can also ignore:

copy_addr [take] %0 to %1

**Comment by gottesmm** [comment]
The reason why we do not want to eliminate retains is we want to make sure that we fail at the actual program termination point, not before.

There is a possibility where we could by removing the retain change a balanced retain count to be unbalanced resulting in us touching a deallocated object.

In contrast, removing a release just makes us leak even more, which we are fine with.

**Comment by jckarter** [comment]
Changing `copy_addr [take] %0 to %1` into a take-initialization is not safe in all cases. If the type being operated on contains weak references or other side references, it will corrupt the weak reference table if you fail to deregister the value currently at `%1`. Maybe it's safe for all currently loadable types, but if we move in the direction of semantic ARC and start representing all types as SSA values, would this be a liability?

Again, `copy_addr [take] [init]` are `load; store` are *always* equivalent, so I don't understand why we're limiting this transformation to unreachable blocks.

**Comment by shajrawi** [comment]
@gottesmm can you please answer @jckarter 's input, I feel you are more qualified to answer his concerns.
I also updated the PR with all the changes you requested.

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Comment by gottesmm** [comment]
Joe, can we talk about this tomorrow in person?

**Comment by gottesmm** [comment]
@jckarter I mean ; )

**Comment by shajrawi** [comment]
Updated the PR: removed code modifications and replaced it with a comment that sums-up the offline discussion with @jckarter and @gottesmm

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Comment by shajrawi** [comment]
@swift-ci Please smoke test

**Review by gottesmm — APPROVED** [review]

**Pull request merged**